eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.Sign/SignArtifacts.cs

```diff
@@ -0,0 +1,178 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.Versioning;
+using Microsoft.Build.Framework;
+
+[assembly:UnsupportedOSPlatform("windows")]
+
+namespace Microsoft.DotNet.SourceBuild.Tasks.Sign;
+
+public class SignArtifacts : Microsoft.Build.Utilities.Task
+{
+    /// <summary>
+    /// List of source-built assets to sign
+    /// </summary>
+    [Required]
+    public required ITaskItem[] SignableSourceBuiltAssets { get; init; }
+
+    /// <summary>
+    /// Path to the eng directory
+    /// </summary>
+    [Required]
+    public required string DotNetEngDirectory { get; init; }
+
+    /// <summary>
+    /// Path to the logs directory
+    /// </summary>
+    [Required]
+    public required string LogsDirectory { get; init; }
+
+    /// <summary>
+    /// Sign-specific properties to pass to the signing process
+    /// </summary>
+    public string SignProperties { get; init; } = string.Empty;
+
+    /// <summary>
+    /// The official build ID
+    /// </summary>
+    public string OfficialBuildId { get; init; } = string.Empty;
+
+    private string SigningPropsTargetPath => Path.Combine(DotNetEngDirectory, Config.SigningPropsFileName);
+    private const string SignBinlogFileName = "sign-source-built-artifacts.binlog";
+    private const int SigningTimeout = 60 * 60 * 1000 * 2; // 2 hours
+
+    public override bool Execute()
+    {
+        try
+        {
+            PrepareForSigning();
+            RunSigning();
+            Log.LogMessage(MessageImportance.High, "Signing complete.");
+        }
+        catch (Exception ex)
+        {
+            Log.LogError($"Signing failed: {ex.Message}");
+        }
+        finally
+        {
+            CleanupSigning();
+        }
+
+        return !Log.HasLoggedErrors;
+    }
+
+    private void PrepareForSigning()
+    {
+        if (!Directory.Exists(LogsDirectory))
+        {
+            Directory.CreateDirectory(LogsDirectory);
+        }
+
+        // Place the Signing.props file into the repo's eng directory for
+        // Arcade's signing infra to find
+        File.Copy(Config.SigningPropsPath, SigningPropsTargetPath, overwrite: true);
+    }
+
+    private void CleanupSigning()
+    {
+        if (File.Exists(SigningPropsTargetPath))
+        {
+            File.Delete(SigningPropsTargetPath);
+        }
+    }
+
+    /// <summary>
+    /// Runs the signing process
+    /// </summary>
+    private void RunSigning()
+    {
+        using (var process = new Process())
```

Comment:
If moved into the UB tasks, you could utilize the ProcessService type being added in https://github.com/dotnet/dotnet/pull/2110